webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+		now := time.Now()
+		// Reset rate limit state if ResetTime has passed
+		if now.After(bw.rateLimit.ResetTime) {
+			bw.rateLimit.RateLimited = false
+			// Discord will tell us the actual number remaining after the next request
+			// but give it at least 1 request so it can make the next request
+			bw.rateLimit.Remaining = 1
+			bw.rateLimit.ResetTime = time.Time{}
+			bw.rateLimit.ResetAfter = 0
+			// logger.Printf("Bucket %s rate limit reset",  bw.rateLimit.BucketID)
+		}
+
+		// We're safe to sleep this particular rate limit bucket without impacting other threads
+		// because if Discord is rate limiting us; any other worker (transient messages)
+		// can't process anyway; Discord would just reject it
+		if bw.rateLimit.RateLimited || bw.rateLimit.Remaining <= 0 {
+			wait := time.Until(bw.rateLimit.ResetTime)
+			bw.rateLimit.mu.Unlock()
+			if wait > 0 {
+				logger.Info(fmt.Sprintf("Bucket %s waiting %v", bw.rateLimit.BucketID, wait))
+				time.Sleep(wait)
+			}
+			continue
+		}
+		bw.rateLimit.mu.Unlock()
+
+		// Grab our next message off the quueue
+		results, err := bw.rdb.BLPop(bw.ctx, 5*time.Second, bw.QueueKey).Result()
+		if err == redis.Nil {
+			continue
+		} else if err != nil {
+			logger.Error(fmt.Sprintf("Error popping from %s: %v", bw.QueueKey, err))
+			continue
+		}
+
+		var msg Message
+		if err := json.Unmarshal([]byte(results[1]), &msg); err != nil {
+			logger.Error(fmt.Sprintf("Unmarshal error: %v, JSON: %s", err, results[1]))
+			continue
+		}
+		msg.MessageNumber = rand.Int()
+
+		// We retry rate limited messages N times instead of re-enqueuing them which can slow down
+		// processing other messages; but doesn't let an unhandled bad message block a queue forever
+		// and if we're getting an error on this message; we would very likely have errors on the remaining
+		// messages anyway
+		for attempts := 0; attempts < maxMsgReattempts && !msg.Discardable; attempts++ {
+			_, err := SendWebhook(rdb, &msg, bw.rateLimit)
+			if err == nil {
+				break
+			}
+
+			logger.Info(fmt.Sprintf("Retry %d for %d in bucket %s: %v", attempts+1, msg.MessageNumber, bw.rateLimit.BucketID, err))
+			if attempts == 2 {
+				logger.Warning(fmt.Sprintf("Dropped after 3 retries: %s", msg.Payload.URL))
+				break
+			}
+			wait := GetWorkerRateLimitSleepTime(bw.rateLimit)
+			if wait > 0 {
+				time.Sleep(wait)
+			} else {
+				// We failed to send for some reason other than rate limiting
+				// back off before attempting again
+				time.Sleep(time.Second * 2 * time.Duration(attempts+1))
+			}
+		}
+	}
+}
+
+func ExecuteWebhook(method string, payload []byte, url string) (*http.Response, error) {
+	req, err := http.NewRequest(method, url, bytes.NewBuffer(payload))
+	if err != nil {
+		return nil, fmt.Errorf("request error: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("send error: %v", err)
+	}
+	defer resp.Body.Close()
```

Comment:
are we sure to never want to read the body? 🤔 I would've expected this call to be done at the calling side of `ExecuteWebhook`, but ok :)